### PR TITLE
Documentation improvements and incidental refactoring

### LINKED ISSUE #750: Guidelines for external packages vs contributions to Hypothesis
Continuing the discussion from #742 so as to not have this on an unrelated issue.

Copying my suggestion from there:

> I don't think that we have a formal policy about strategies. I think roughly the following are the guidelines I've been using:
>
> Either an external package or adding it to Hypothesis is fine for most strategies.
> * If the strategy is likely to be of widespread use, we should encourage adding it to Hypothesis core (and offer to help with the work, as this isn't totally straightforward).
> * If the strategy is likely to be of quite niche use, there's still no problem adding it to Hypothesis core, but we don't need to be quite so enthusiastic about doing so.
> (Plus when adding something to Hypothesis core usual rules about things needing an external dependency going in as hypothesis.extra packages)
> Roughly the trade-off is that it's much easier to get things working outside of Hypothesis because there's more freedom to experiment and fewer requirements. but being inside Hypothesis gets you access to somewhat cooler implementation features and better long-term guarantees about maintenance.

@alexwlchan and @mulkieran I think both felt that this was too far in the direction of include by default I think. Is that a fair characterisation? In particular https://github.com/HypothesisWorks/hypothesis-python/issues/742#issuecomment-318768634 and https://github.com/HypothesisWorks/hypothesis-python/issues/742#issuecomment-318663496

My main reasoning as to why I still think an include by default policy is better is that it stops people who are developing things on top of Hypothesis from having to worry about compatibility with different versions of Hypothesis, and it stops us from worrying about breaking their libraries (as these often tend to use slightly more than the public API). In particular the .extra packages exist in Hypothesis because historically I found keeping multiple versions in sync very annoying.

OTOH the public API is significantly more capable than it was when I merged the .extra packages in, so maybe that's not a big deal? Maybe we *should* be trying to divest more, and just make sure that people can do so entirely through the public API.

### LINKED ISSUE #1109: Document necessary and sufficient conditions for version number bumps
Hypothesis's versioning scheme is "semver", but what semver actually means is horribly underspecified and whether something is a breaking change, new feature, etc. is pretty arguable. We also often bump the minor version in cases that arguably we could have got away with a patch version.

I don't expect us to ever be able to fully automatically decide this, but it would be nice to define some conditions where we know the correct answer and can just say "This is a minor release because X" or "This would require a major release because Y" rather than it *always* being a judgement call.

For example, the following would be reasonable rules:

* The signature of functions in the public API may not change in patch releases (e.g. the question of versioning that came up in #1106).
* New health checks may not be added in patch releases
* New deprecations may not be added in patch releases

(This isn't intended to be exhaustive in any way, just some initial thoughts)

This should be part of the public documentation around versioning, and we should add something to the review guide about checking it.

### PR DESCRIPTION
**Pull request by Zac-HD**
This pull has been extracted from #1129.

It consists of a variety of improvements to documentation, mostly but not exclusively for contributors (closes #1109 and closes #750).

I have also included some incidental refactoring here, so that #1131 can focus entirely on the `characters()` changes.

### CODE REVIEW

**Comment by Zac-HD** [comment]
More things to do:

- The explicit reprs for MappedSearchStrategy subclasses should show how the strategy is constructed - leave the info hiding for `@defines_strategy`, ie `characters(...).map(u''.join)` instead of `StringStrategy()`.  Also add caching.
- In related news, the repr for `one_of(a, b).map(...)` is wrong: you get `a | b.map(...)`.  This can and should be fixed by changing the `one_of` repr.

File: guides/documentation.rst [hunk 1]
```diff
@@ -57,11 +57,27 @@ where every pull request that touches ``./src`` results in a new release.
...
-the repository root, and:
+the repository root.  The first line of the file specifies the component
+of the version number that will be updated, according to our
+`semantic versioning <http://semver.org/>`_ policy.
+
+- ``RELEASE_TYPE: major`` is for breaking changes, and will only be used by the
+  core team after extensive discussion.
+- ``RELEASE_TYPE: minor`` is for anything that adds to the public (ie documented)
+  API, changes an argument signature, or adds a new deprecation or health check.
+  Minor (or patch) releases **may not** cause errors in any code that runs
```
[the remark this answers is not in the document]
Comment (dchudz) [reply guides/documentation.rst]: Do we have any guidance about what kinds of behavior changes are acceptable for a minor release (say, in cases where neither the old nor new behavior is an exception)? 

It's okay not to say anything about that if we don't know what we would say, but this is a good place for that if we have it.

[the remark this answers is not in the document]
Comment (Zac-HD) [reply guides/documentation.rst]: Whoops, I'd missed that.  

In principle any non-breaking change is allowed in a minor release; in practice it's also limited by our various style guides and the scope of the library.  I think it's better not to write more here.

**Comment by Zac-HD** [comment]
:tada:

@DRMacIver, if you want any changes let me know soon - otherwise merge it, or I'll merge it early on Sunday (UTC).

**Pull request merged**